proxy/round_tripper/proxy_round_tripper.go

```diff
@@ -255,15 +286,26 @@ func (rt *roundTripper) RoundTrip(originalRequest *http.Request) (*http.Response
 		originalRequest.Body.Close()
 	}
 
-	finalErr := err
-	if finalErr == nil {
-		finalErr = selectEndpointErr
+	// If we have an error from the round trip, we prefer it over errors
+	// returned from selecting the endpoint, see declaration of
+	// selectEndpointErr for details.
+	if err == nil {
+		err = selectEndpointErr
 	}
 
-	if finalErr != nil {
-		rt.errorHandler.HandleError(reqInfo.ProxyResponseWriter, finalErr)
-		return nil, finalErr
+	if err != nil {
+		rt.errorHandler.HandleError(reqInfo.ProxyResponseWriter, err)
+		return nil, err
 	}
+
+	// Record the times from the last attempt, but only if it succeeded.
+	reqInfo.DnsStartedAt = time.Unix(0, trace.tDNSStart.Load())
```

Comment:
I'd prefer having this in the requestTracer type.
e.g.
```
reqInfo.DnsStartedAt = trace.DnsStartedAt()
```
and remove the `trace.DnsTime()` etc. functions as we no longer need them since we only record timestamps now.

Reply:
We are still using those for the error logs, so I will keep them.